### PR TITLE
[AndroidCrypto] Implement IExportPal for X509 certs

### PR DESCRIPTION
**Pull request by elinor-fung**
cc @jkoritzinsky @steveisok @AaronRobinsonMSFT @bartonjs

### CODE REVIEW

File: src/libraries/Common/src/System/Security/Cryptography/RSAAndroid.cs [hunk 1]
```diff
@@ -32,7 +32,7 @@ public RSAAndroid(int keySize)
 
             internal RSAAndroid(SafeRsaHandle key)
             {
-                _key = new Lazy<SafeRsaHandle>(key);
+                _key = new Lazy<SafeRsaHandle>(key.DuplicateHandle());
```
Comment (bartonjs) [inline src/libraries/Common/src/System/Security/Cryptography/RSAAndroid.cs:35]: Are there any callers that were previously using DuplicateKey that should stop doing that now so we don't finalize an intermediate value?

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c [hunk 2]
```diff
@@ -153,10 +154,52 @@ int32_t AndroidCryptoNative_X509DecodeCollection(const uint8_t *buf, int32_t buf
...
+    }
+
+    // CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+    loc[certType] = JSTRING("X.509");
+    loc[certFactory] = (*env)->CallStaticObjectMethod(env, g_CertFactoryClass, g_CertFactoryGetInstance, loc[certType]);
+    ON_EXCEPTION_PRINT_AND_GOTO(cleanup);
+
+    // CertPath certPath = certFactory.generateCertPath(certList);
+    // byte[] encoded = certPath.getEncoded("PKCS7");
+    loc[certPath] = (*env)->CallObjectMethod(env, loc[certFactory], g_CertFactoryGenerateCertPathFromList, loc[certList]);
```
Comment (bartonjs) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:186]: Just confirming: This works for unrelated certificates?  It hasn't actually tried building the path yet?

File: src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.ExportPal.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                        algorithm = new ECDsaImplementation.ECDsaAndroid(ec);
+                        break;
+                    case SafeRsaHandle rsa:
+                        algorithm = new RSAImplementation.RSAAndroid(rsa);
+                        break;
+                    default:
+                        throw new NotSupportedException(SR.NotSupported_KeyAlgorithm);
+                }
+
+                return algorithm.ExportEncryptedPkcs8PrivateKey(password, s_windowsPbe);
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.ExportPal.cs:64]:
```suggestion
                using (algorithm)
                {
                    return algorithm.ExportEncryptedPkcs8PrivateKey(password, s_windowsPbe);
                }
```

File: src/libraries/Common/src/System/Security/Cryptography/RSAAndroid.cs (hunk 1 above)
Comment (elinor-fung) [reply src/libraries/Common/src/System/Security/Cryptography/RSAAndroid.cs:35]: No, this was a bug from integrating the certs with RSA.

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c (hunk 2 above)
Comment (elinor-fung) [reply src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:186]: Going through `CertificateFactory` (this) works - it doesn't try building the path yet and isn't validated unless put through a `CertPathValidator`. Going through `CertPathBuilder` would not work, as it does do the path building.

**Review by bartonjs — APPROVED** [review]

File: src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.X509.cs [hunk 4]
```diff
@@ -10,23 +10,33 @@ internal static partial class Interop
...
         {
-            return GetDynamicBuffer((ptr, buf, i) => X509Encode(ptr, buf, i), x);
+            int len = 0;
+            int ret = X509Encode(x, null, ref len);
+            if (ret != INSUFFICIENT_BUFFER)
+                throw new CryptographicException();
+
+            byte[] encoded = new byte[len];
+            ret = X509Encode(x, encoded, ref len);
+            if (ret != SUCCESS)
```
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.X509.cs:30]: I assume the answer is "no", but is there any chance len here might != encoded.Length, i.e. on the previous call we got one length number but here we got a different one?

Comment (elinor-fung) [reply src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.X509.cs:30]: They should be the same (as long as `ret == SUCCESS`). The properties of the object being encoded can't be changed so the encoded length wouldn't change between the previous call and this one.

**Pull request merged**